### ISSUE TITLE
Docs enhancement: example of providing a quoted arg

### ISSUE THREAD

**Issue by nick-youngblut** [markfairbanks/tidytable#57]
The tidytable docs include some examples of providing a non-quoted function arg, such as:

```
library(rlang)

df <- data.table(x = c(1,1,1), y = c(1,1,1), z = c("a","a","b"))

add_one <- function(.data, add_col) {
  add_col <- enexpr(add_col)

  .data %>%
    mutate.(new_col = !!add_col + 1)
}

df %>%
  add_one(x)
```

... but the docs don't have an example of if the user provides a quoted arg, such as:

```
make_subsets = function(to_filter, dt){
    dt %>% dt_filter(Species == to_filter)
}
iris$Species %>% unique %>% as.list %>% lapply(make_subsets, dt=as.data.table(iris))
```

**Comment by nick-youngblut** [comment]
To be a bit more specific, it would help to know why this works:

```
iris$Species %>% unique %>% as.list %>% lapply(make_subsets, dt=as.data.table(iris))
```

...and this doens't:

```
c('setosa', 'versicolor', 'virginica') %>% as.list %>% lapply(make_subsets, dt=as.data.table(iris))
```

...with the `make_subsets()` function shown above, and how to deal with the latter.

**Comment by markfairbanks** [comment]
Ah yep, thanks for opening this issue. I actually just discovered this the other day.

This is a side effect of how `tidytable` uses `eval(expr(dt_expression)` at the end of all code to build the data.table expression.

Even though "to_filter" is a normal variable input, it still needs to be quoted/unquoted.

```r
make_subsets = function(to_filter, dt){
  to_filter <- enexpr(to_filter)
  
  dt %>% dt_filter(Species == !!to_filter)
}
```

I've been debating how to add this to the documentation, as this is a behavior that differs from the tidyverse

**Comment by nick-youngblut** [comment]
It would also be nice to have an example of dynamically adding a new column with `dt_mutate` or dynamically filtering. For example:

```
df <- data.table(x = c(1,1,1), y = c(1,1,1), z = c("a","a","b"))

add_one <- function(.data, add_col, new_col) {
  add_col <- enexpr(add_col)
  new_col <- enexpr(add_col)

  .data %>%
    mutate.(!!new_col = !!add_col + 1)
}

df %>%
  add_one(x)
```

...and for filter:

```
df <- data.table(x = c(1,1,1), y = c(1,1,1), z = c("a","a","b"))

dynamic_filter <- function(.data, to_filter, filter_column) {
  to_filter <- enexpr(to_filter)
  filter_column <- enexpr(filter_column)

  .data %>%
    filter.(!!filter_column == !!to_filter)
}

df %>%
  dynamic_filter(x, a, z)
```

Both of these examples don't work

**Comment by nick-youngblut** [comment]
> Even though "to_filter" is a normal variable input, it still needs to be quoted/unquoted.

Is there a why to allow the function to work regardless of whether the user quotes or does not quote the variable? For example:

```
df <- data.table(x = c(1,1,1), y = c(1,1,1), z = c("a","a","b"))

add_one <- function(.data, add_col) {
  add_col <- enexpr(add_col)
  .data %>%
    mutate.(new_col = !!add_col + 1)
}

df %>%  add_one(x)   # workds
df %>%  add_one('x')   # fails with code shown above
```

**Comment by markfairbanks** [comment]
I think I'll just build a vignette for these, but for your mutate example this works just like it does in the tidyverse when naming a new column:

```r
df <- data.table(x = c(1,1,1), y = c(1,1,1), z = c("a","a","b"))

add_one <- function(.data, add_col, new_col) {
  add_col <- enexpr(add_col)
  new_col <- enexpr(add_col)

  .data %>%
    mutate.(!!new_col := !!add_col + 1)
}

df %>%
  add_one(x)
```

And your filter example works like so - I think you just had some arguments mixed up:

```r
df <- data.table(x = c(1,1,1), y = c(1,1,1), z = c("a","a","b"))

dynamic_filter <- function(.data, to_filter, filter_column) {
  to_filter <- enexpr(to_filter)
  filter_column <- enexpr(filter_column)

  .data %>%
    filter.(!!filter_column == !!to_filter)
}

df %>%
  dynamic_filter("a", z)
```

**Comment by markfairbanks** [comment]
> Is there a why to allow the function to work regardless of whether the user quotes or does not quote the variable?

This behavior unfortunately can't be changed. This is pretty core to how data.table runs

**Comment by nick-youngblut** [comment]
Thanks for all of the quick replies and explanations! I'm still trying to figure out the following:

```
make_subsets = function(to_filter, filter_col, dt){
    to_filter = enexpr(to_filter)
    filter_col = enexpr(filter_col)
    dt_filter(dt, !!filter_col == !!to_filter)
}
iris$Species %>% unique %>% as.list %>% 
    lapply(make_subsets, filter_col=Species, dt=as.data.table(iris))   # works

c('setosa', 'versicolor', 'virginica') %>% as.list %>% 
    lapply(make_subsets, filter_col=Species, dt=as.data.table(iris))   # doesn't work

c('setosa', 'versicolor', 'virginica') %>% as.factor %>% as.list %>% 
    lapply(make_subsets, filter_col=Species, dt=as.data.table(iris))   # works
```

...but the `as.factor %>% as.list %>% ` doesn't work for everything:

```
df <- data.table(x = c(1,1,1), y = c(1,1,1), z = c("a","a","b"))

add_one <- function(add_col, .data) {
  add_col <- enexpr(add_col)

  .data %>%
    mutate.(new_col := !!add_col + 1)
}

c('x') %>% as.factor %>% as.list %>%
    lapply(add_one, .data=df)  # generates NA's
```

**Comment by markfairbanks** [comment]
> Thanks for all of the quick replies and explanations!

Yep, no problem. I was actually just working on some R stuff so the timing worked out.

> I'm still trying to figure out the following:

So there's a couple things here:

1. `lapply()` can iterate over a vector, so you can skip using `as.list()`. `lapply()` is named that because it _returns_ a list, not that it needs a list input.
2. If you're ever doing filtering of a data.frame/data.table, I would recommend working on character columns instead of factors. So I added a conversion step to the example.

```r
make_subsets = function(to_filter, filter_col, dt){
    to_filter = enexpr(to_filter)
    filter_col = enexpr(filter_col)
    dt_filter(dt, !!filter_col == !!to_filter)
}

# Convert two a character first
iris_df <- iris %>% mutate_across.(is.factor, as.character)

c('setosa', 'versicolor', 'virginica') %>%
    lapply(make_subsets, filter_col=Species, dt = iris_df)
```

**Comment by nick-youngblut** [comment]
> I would recommend working on character columns instead of factors

Is that due to a speedup when using characters instead of factors, or can the user get unintended results when using factors?

Is there a way to pass an unquoted arg trough multiple layers of functions, such as this trivial (but non-functioning) example:

```
df <- data.table(x = c(1,1,1), y = c(1,1,1), z = c("a","a","b"))

.add_one <- function(.data, add_col) {
  add_col <- enexpr(add_col)
  .data %>%
    mutate.(new_col = !!add_col + 1)
}

add_one <- function(.data, add_col) {
  .add_one(.data, add_col)
}

df %>%
  add_one(x)  # doesn't work

df %>%
  .add_one(x)  # does work
```

**Comment by markfairbanks** [comment]
> Is that due to a speedup when using characters instead of factors, or can the user get unintended results when using factors?

It's because the user can get unintended results. I've never seen the issue, but supposedly it can occur.

> Is there a way to pass an unquoted arg through multiple layers of functions, such as this trivial (but non-functioning) example:

Actually this is a great question. This is possible using `rlang`, but can't be recreated using base R. This example is one of the biggest reasons `rlang` is such a great package, and is the reason I used it to build tidytable.

The basic idea is you keep quoting/unquoting for the entire chain of functions:

```r
df <- data.table(x = c(1,1,1), y = c(1,1,1), z = c("a","a","b"))

.add_one <- function(.data, add_col) {
  add_col <- enexpr(add_col)
  .data %>%
    mutate.(new_col = !!add_col + 1)
}

add_one <- function(.data, add_col) {
  add_col <- enexpr(add_col)
  
  .add_one(.data, !!add_col)
}

df %>%
  add_one(x)  # now works
```

**Comment by nick-youngblut** [comment]
Thanks @markfairbanks for all of the help today! I feel that I now understand much more about how to write functions that use tidytable!

**Comment by markfairbanks** [comment]
Yep, no problem! If you run into any more questions feel free to reach out.

As for this issue, I'll leave it open until I create a vignette to help with building functions.

**Comment by nick-youngblut** [comment]
I am having a problem with unquoted args with `furrr`:

```
make_subsets = function(to_filter, filter_col, dt){
    to_filter = enexpr(to_filter)
    filter_col = enexpr(filter_col)
    dt_filter(dt, !!filter_col == !!to_filter)
}

# Convert factor cols to character cols first
iris_df <- iris %>% mutate_across.(is.factor, as.character)

#-- works --#
c('setosa', 'versicolor', 'virginica') %>%
    purrr::map(make_subsets, filter_col=Species, dt=iris)

#-- doesn't work --#
c('setosa', 'versicolor', 'virginica') %>%
    furrr::future_map(make_subsets, filter_col=Species, dt=iris)
```

The same error occurs when using `future.apply::future_lapply` 

Any ideas? Maybe I should post something on the `furrr` or `future` repo?

**Comment by nick-youngblut** [comment]
`plyr` + `doParallel` also seems to have problems:

```
make_subsets = function(to_filter, filter_col, dt){
    to_filter = enexpr(to_filter)
    filter_col = enexpr(filter_col)
    dt_filter(dt, !!filter_col == !!to_filter)
}

# Convert factor cols to character cols first
iris_df <- iris %>% mutate_across.(is.factor, as.character)

# -- works -- #
c('setosa', 'versicolor', 'virginica') %>%
   plyr::llply(make_subsets, filter_col=Species, dt=iris, .parallel=FALSE)

# -- doesn't work -- #
doParallel::registerDoParallel(2)
c('setosa', 'versicolor', 'virginica') %>%
   plyr::llply(make_subsets, filter_col=Species, dt=iris, .parallel=TRUE)
```

**Comment by nick-youngblut** [comment]
Also `%in%` doesn't work for `enexpr`'d variables, but `==` does:

```
make_subsets = function(to_filter, filter_col, dt){
    to_filter = enexpr(to_filter)
    filter_col = enexpr(filter_col)
    print(filter_col == 'Species')                           # prints TRUE
    print(filter_col %in% c('Species', 'Genus'))    # throws an error
}

# Convert factor cols to character cols first
iris_df <- iris %>% mutate_across.(is.factor, as.character)

c('setosa', 'versicolor', 'virginica') %>%
   lapply(make_subsets, filter_col=Species, dt=iris)

**Comment by markfairbanks** [comment]
> I am having a problem with unquoted args with furrr

This appears to be a furrr issue. Might be worth opening an issue on their page using the tidyverse (I wouldn't use tidytable because it's less common).

This is the example I would use

```r
library(tidyverse)

make_subsets = function(to_filter, filter_col, dt){
    filter_col <- enexpr(filter_col)
    
    filter(dt, !!filter_col == to_filter)
}

# Convert factor cols to character cols first
iris_df <- iris %>% mutate_if(is.factor, as.character)

#-- works --#
c('setosa', 'versicolor', 'virginica') %>%
    purrr::map(make_subsets, filter_col=Species, dt=iris_df)

#-- doesn't work --#
c('setosa', 'versicolor', 'virginica') %>%
    furrr::future_map(make_subsets, filter_col=Species, dt=iris_df)
```

**Comment by markfairbanks** [comment]
> Also %in% doesn't work for enexpr'd variables, but == does:

The second one fails because you're comparing a symbol to a character. `==` works (even though it probably shouldn't), but it's not designed for that comparison.

This will work as intended.

```r
make_subsets = function(to_filter, filter_col, dt){
    to_filter = enexpr(to_filter)
    filter_col = enexpr(filter_col)
    
    filter_col_char = as.character(filter_col)

    print(filter_col_char == 'Species')
    print(filter_col_char %in% c('Species', 'Genus'))
}

# Convert factor cols to character cols first
iris_df <- iris %>% mutate_across.(is.factor, as.character)

c('setosa', 'versicolor', 'virginica') %>%
   lapply(make_subsets, filter_col=Species, dt=iris)
```

**Comment by nick-youngblut** [comment]
In regards to the parallelization issue, I posted it on the dplyr repo, and Hadley Wickham shut it down: https://github.com/tidyverse/dplyr/issues/5135

It seems that tidytable is limited is its ability to be used in functions that are run in parallel.

**Comment by markfairbanks** [comment]
Interesting. Well, it is what it is. Normally there are ways to set up your processes so this isn't an issue. For example:

```
iris %>%
  group_split.(Species)
```

Then you can put each data frame in a parallel operation from there.

**Comment by markfairbanks** [comment]
Your original example now works in the developmental version. You no longer need to quote/unquote function arguments that are not column names.

```r
library(data.table, warn.conflicts = FALSE)
library(tidytable, warn.conflicts = FALSE)

make_subsets = function(to_filter, dt) {
    dt %>% filter.(Species == to_filter)
}

c('setosa', 'versicolor', 'virginica') %>%
  lapply(make_subsets, dt=as.data.table(iris))
```

**Comment by markfairbanks** [comment]
All custom functions can be created using the same syntax that you use in the tidyverse. [This blog post](https://www.tidyverse.org/blog/2019/06/rlang-0-4-0/) also talks about the embracing pattern `{{}}` that can now be used in tidytable.

**Comment by nick-youngblut** [comment]
>I think I'll just build a vignette for these

Any updated on creating a vignette for this? I keep running into issues, especially when providing quoted variables to the function. For instance, neither of these examples work:

```
dynamic_mean <- function(dt, num_col, group_col){
  dt %>%
    summarize.(new_val = mean({{num_col}}, na.rm=TRUE),
               .by = c({{group_col}}))
}

iris %>% dynamic_mean('Sepal.Length', 'Species')
```


```
dynamic_mean <- function(dt, num_col, group_col){
  num_col <- enexpr(num_col)
  group_col  <- enexpr(group_col)
  dt %>%
    summarize.(new_val = mean(!!num_col, na.rm=TRUE),
               .by = c(!!group_col))
}

iris %>% dynamic_mean('Sepal.Length', 'Species')
```

I'm using `tidytable 0.9.0`

**Comment by markfairbanks** [comment]
Also worth noting - when you originally asked this question a couple years ago there were some oddities to how tidytable worked with quosures. Those are now fixed. So you should be able to use tidy evaluation just like you do with dplyr.

**Comment by markfairbanks** [comment]
Another tip - in a "column selection" context (like in `select()` or using `.by`) you can pass vectors of strings with the `all_of()` helper.

``` r
pacman::p_load(tidytable)

dynamic_mean <- function(dt, num_col, group_cols){
  dt %>%
    summarize.(new_val = mean(.data[[num_col]], na.rm=TRUE),
               .by = all_of(group_cols))
}

df <- tidytable(x = c("a", "a", "b"), y = c("a", "a", "b"), z = 1:3)

df %>% dynamic_mean("z", c("x", "y"))
#> # A tidytable: 2 × 3
#>   x     y     new_val
#>   <chr> <chr>   <dbl>
#> 1 a     a         1.5
#> 2 b     b         3
```
